providers/common/bio_prov.c

```diff
@@ -94,3 +109,110 @@ int ossl_prov_bio_printf(BIO *bio, const char *format, ...)
     return ret;
 }
 
+#ifndef FIPS_MODULE
+
+/* No direct BIO support in the FIPS module */
+
+static int bio_core_read_ex(BIO *bio, char *data, size_t data_len,
+                            size_t *bytes_read)
+{
+    return ossl_prov_bio_read_ex(BIO_get_data(bio), data, data_len, bytes_read);
+}
+
+static int bio_core_write_ex(BIO *bio, const char *data, size_t data_len,
+                             size_t *written)
+{
+    return ossl_prov_bio_write_ex(BIO_get_data(bio), data, data_len, written);
+}
+
+static long bio_core_ctrl(BIO *bio, int cmd, long num, void *ptr)
+{
+    /* We don't support this */
+    assert(0);
+    return 0;
+}
+
+static int bio_core_gets(BIO *bio, char *buf, int size)
+{
+    /* We don't support this */
+    assert(0);
+    return -1;
+}
+
+static int bio_core_puts(BIO *bio, const char *str)
+{
+    /* We don't support this */
+    assert(0);
+    return -1;
+}
+
+static int bio_core_new(BIO *bio)
+{
+    BIO_set_init(bio, 1);
+
+    return 1;
+}
+
+static int bio_core_free(BIO *bio)
+{
+    BIO_set_init(bio, 0);
+
+    return 1;
+}
+
+typedef struct bio_prov_global_st {
+    BIO_METHOD *biocoremeth;
+} BIO_PROV_GLOBAL;
+
+static void *bio_prov_ossl_ctx_new(OPENSSL_CTX *libctx)
+{
+    BIO_PROV_GLOBAL *bpgbl = OPENSSL_zalloc(sizeof(*bpgbl));
+    BIO_METHOD *biocoremeth = NULL;
+
+    biocoremeth = BIO_meth_new(BIO_TYPE_CORE_TO_PROV, "BIO to Core filter");
+    if (biocoremeth == NULL
+            || !BIO_meth_set_write_ex(biocoremeth, bio_core_write_ex)
+            || !BIO_meth_set_read_ex(biocoremeth, bio_core_read_ex)
+            || !BIO_meth_set_puts(biocoremeth, bio_core_puts)
+            || !BIO_meth_set_gets(biocoremeth, bio_core_gets)
+            || !BIO_meth_set_ctrl(biocoremeth, bio_core_ctrl)
+            || !BIO_meth_set_create(biocoremeth, bio_core_new)
+            || !BIO_meth_set_destroy(biocoremeth, bio_core_free)) {
+        OPENSSL_free(bpgbl);
+        BIO_meth_free(biocoremeth);
+        return NULL;
+    }
+
+    bpgbl->biocoremeth = biocoremeth;
+
+    return bpgbl;
+}
+
+static void bio_prov_ossl_ctx_free(void *vbpgbl)
+{
+    BIO_PROV_GLOBAL *bpgbl = vbpgbl;
+
+    BIO_meth_free(bpgbl->biocoremeth);
+    OPENSSL_free(bpgbl);
+}
+
+static const OPENSSL_CTX_METHOD bio_prov_ossl_ctx_method = {
+    bio_prov_ossl_ctx_new,
+    bio_prov_ossl_ctx_free,
+};
+
+BIO *BIO_new_from_core_bio(OPENSSL_CTX *libctx, OSSL_CORE_BIO *corebio)
+{
+    BIO_PROV_GLOBAL *bpgbl = openssl_ctx_get_data(libctx,
+                                                  OPENSSL_CTX_BIO_PROV_INDEX,
+                                                  &bio_prov_ossl_ctx_method);
```

Comment:
Hmmmm, do you really expect the code to deliver several core OSSL_DISPATCH (the *in* argument to the provider init function)?  If not, I'm not sure I understand why you can't have just a single provider global `bio_prov_ossl_ctx_method` that you populate once.

If you really want to tie it to the provider context, you could simply add a BIO method to the provider context itself instead of (unnecessarily) going through the library context.  Just make it an actual structure:

``` C
struct provctx_st {
    OPENSSL_CTX *libctx;
    BIO_METHOD *biocoremeth;
}
```

Reply:
> Hmmmm, do you really expect the code to deliver several core OSSL_DISPATCH (the in argument to the provider init function)? If not, I'm not sure I understand why you can't have just a single provider global bio_prov_ossl_ctx_method that you populate once.

We've gone around this loop (several times) before. Global variables are problematic in the providers, because you have to figure out how to free them again when the provider unloads (ref-counting does not work). This is a minefield. Lets not go around that loop again.

> If you really want to tie it to the provider context, you could simply add a BIO method to the provider context itself instead of (unnecessarily) going through the library context. Just make it an actual structure:

Its simple to say, but not so simple to do. Do-able - but introducing a provider context object which isn't just the libctx is probably a whole PR in itself. We assume all over the place that the provctx is the library context. Now you could just put the libctx first in the provctx structure - and that assumption would still hold. But it is really ugly. I really don't want to go down that path in this PR.

Reply:
> We assume all over the place that the provctx is the library context.

I sure hope that *we do not*.  There is a central spot where the provider context would be defined, `providers/common/include/prov/provider_ctx.h`, and there is a macro to extract the library context from the provider context - `PROV_LIBRARY_CONTEXT_OF`, *exactly so we can avoid that kind of assumption*.  Why do you think I pushed for that macro in the first place?

Reply:
Ah, I found *one spot* where that kind of assumption is made:

https://github.com/openssl/openssl/blob/8c30dfee3ea038b71f339f193331ed6ac11e3055/providers/fips/fipsprov.c#L652

Reply:
Have a look at #11803

Reply:
Ok, I rebased this on top of #11803